pageserver/src/tenant/tasks.rs

```diff
@@ -553,6 +559,65 @@ pub(crate) async fn delay_by_lease_length(
     }
 }
 
+struct Iteration {
+    started_at: Instant,
+    period: Duration,
+    kind: BackgroundLoopKind,
+}
+
+struct IterationResult<O> {
+    output: O,
+    elapsed: Duration,
+}
+
+impl Iteration {
+    #[instrument(skip_all)]
+    pub(crate) async fn run<Fut, O>(self, fut: Fut) -> IterationResult<O>
+    where
+        Fut: std::future::Future<Output = O>,
+    {
+        let Self {
+            started_at,
+            period,
+            kind,
+        } = self;
+
+        let mut fut = std::pin::pin!(fut);
+
+        // Wrap `fut` into a future that logs a message every `period` so that we get a
+        // very obvious breadcrumb in the logs _while_ a slow iteration is happening.
+        let liveness_logger = async move {
+            let Some(mut next_warn_at) = self.started_at.checked_add(period) else {
+                warn!("checked_add overflow");
+                return fut.await;
+            };
+            loop {
+                match tokio::time::timeout_at(next_warn_at.into(), &mut fut).await {
+                    Ok(x) => return x,
+                    Err(_) => {
+                        // info level as per the same rationale why warn_when_period_overrun is info
+                        // =>  https://github.com/neondatabase/neon/pull/5724
+                        info!("still running");
+                        next_warn_at = match Instant::now().checked_add(period) {
+                            Some(x) => x,
+                            None => {
+                                warn!("checked_add overflow");
+                                return fut.await;
+                            }
+                        };
+                    }
+                }
+            }
```

Comment:
Why use `timeout_at` over `timeout`?

I was initially thinking there was some wisdom I always fail to see, but now I think it's just extra which causes branch here and the other with `next_warn_at`. tokio will internally handle overflowing this add as "far_future" timer.

So without this manual stuff, we'd be:

```suggestion
            loop {
                match tokio::time::timeout(period, &mut fut).await {
                    Ok(x) => return x,
                    Err(_) => {
                        // info level as per the same rationale why warn_when_period_overrun is info
                        // =>  https://github.com/neondatabase/neon/pull/5724
                        info!("still running");
                    }
                }
            }
```

Reply:
thanks, applied!